rfcs/20200817-jit-code-generation-for-aarch64/README.md

```diff
@@ -0,0 +1,256 @@
+# Proposal to Introduce JIT code generation for AArch64
+
+## Introduction
+
+In the x64 implementation of oneDNN, the execution code is generated at runtime 
+considering the available instruction set, such as AVX512/AVX2/AVX/SSE, 
+and runtime parameters. oneDNN uses JIT assembler (Xbyak) to generate execution 
+code. For AArch64 architecture, we has developed 
+[Xbyak_aarch64](https://github.com/fujitsu/xbyak_aarch64), 
+a port of Xbyak for AArch64. Xbyak_aarch64 provides the same functionality 
+for Aarch64 as Xbyak provides for x64. Therefore, the same functionality that 
+oneDNN achieves for x64 using Xbyak can be achieved for AArch64 from scratch 
+using Xbyak_aarch64.
+**This RFC introduces Xbyak_aarch64 into oneDNN and enables JIT code generation 
+for AArch64.**
+
+Ideally, all implementations written for x64 should be rewritten for AArch64 
+using Xbyak_aarch64 (Fig. Direct method) to obtain maximum performance of 
+AArch64.
+
+![Direct method](direct_method.png)
+
+However, this process takes a lot of man-hours. To solve this problem, 
+we have developed 
+[Xbyak_translator_aarch64](https://github.com/fujitsu/xbyak_translator_aarch64). 
+By using Xbyak_translator_aarch64 instead of Xbyak in JIT code generation 
+for x64, we can generate JIT code for AArch64 instead of JIT code for x64 
+(Fig. Indirect method). Since we can reuse jit_\*.(cpp|hpp) for x64 with just 
+a few changes for AArch64, and we can reduce our development effort for 
+AArch64 to a great extent. We call this as "Indirect method". So **this RFC 
+also introduces Xbyak_translator_aarch64 and 
+[xed](https://github.com/fujitsu/xed "xed") for AArch64.**
+
+![Indirect method](indirect_method.png)
+
+
+As a third method, it is also possible that we use x64's jit_\*.(cpp|hpp) 
+files and their translation to AArch64 as a baseline, and partially rewrite 
+them from Xbyak functions to Xbyak_aarch64 ones to achieve better performance 
+for AArch64. We call this method as "Mixed-method".
+
+In summary, this RFC allows us to generate JIT code for AArch64 in the 
+following three ways.
+- Direct method:Implement JIT code generation from scratch using Xbyak_aarch64
+- Indirect method:Generate JIT code for AArch64 from jit_\*.(cpp|hpp) 
+for x64 by Xbyak_translator_aarch64
+- Mixed method:Use jit_\*.(cpp|hpp) files for x64 as a baseline, tune them 
+partially with Xbyak_aarch64
+
+We applied these three methods to oneDNN code in v0.21 and confirms that 
+the JIT code is generated on AArch64, it passes tests/gtests/test_\*, 
+and it accelerates oneDNN processing as shows in the slide below.
+The source code is disclosed as 
+[https://github.com/fujitsu/dnnl_aarch64](https://github.com/fujitsu/dnnl_aarch64).
```

Comment:
Could you please elaborate a little more on what kind of implementations you are going to promote? Would it be more like a translated (and tuned) former x64 code or do you plan to have some kind of automation / periodic updates for translated kernels based on new x64 jitted kernels?

Reply:
Currently, almost our implementation was made by the indirect method i.e. automatic translation from x64. Only CNN is optimized by manual assemble code.
We think our translator is useful for supporting future new CPUs (e.g. SVE2, 3, 4...). Then, we opensourced it. If some vendor will get SVE2, we can collaborate them.
For SVE1 (our A64FX cpu), I don't think we need periodic updates by the translator.  
However, we are also researching more optimization of the translator. If  we will get
better number, updating sve1 code is an option of course.

Reply:
In other words, our periodic updates will be made by manual assemble for the time being. I suspect.

Reply:
I see. Good to know. Thank you!